fix: property extractor

jbristow opened the pull request:
Some error messages were falling through because they were lacking exception info, which caused us to miss some errors getting properties.

This diff fixes the properties errors I could see locally and attempts to unify all logging to conform to python best practices and also to highlight potential data loss more explicitly.

ccloes commented:
@stuartio can you review this please?  We found a bug and this is addressing that.

stuartio commented:
Can you give me more context on the bug, and the fix. This PR is a bit of a beast

jbristow commented:
> Can you give me more context on the bug, and the fix. This PR is a bit of a beast

We started noticing that some AkamaiProperty nodes seemed to be missing.  This led to us discovering that it was throwing an error but not being picked up by out logging systems.

Once I fixed that error, I was noticing a lot KeyErrors occurring when trying to get data from dictionaries when the key was not present, so I also fixed those.

While waiting for my local test runs to complete, I also went through all of the logging entries and made sure that things that had exceptions signifying potential data loss or incomplete processing were using `logger.exception()`.  For things that did not prevent grabbing other things, I downgraded them to a "warn" if necessary.

Then I added some tests, and added a linter (ruff) and applied most of the "auto-fixable" fixes and a handful of ones that worried me (things like "variable could be used before declared")

I attempted to keep logic changes limited to anything other than `property`/`property_client`, only applying cosmetic or simplifying changes to other pipelines.

jbristow commented on nodestream_akamai/akamai_utils/client.py:
```diff
@@ -121,22 +131,13 @@ def _post_api_from_relative_path(self, path, body, params=None, headers=None):
-                continue
-
-            yield elem
-            seen.add(k)
```
unused

jbristow commented on nodestream_akamai/akamai_utils/client.py:
```diff
@@ -85,9 +90,13 @@ def _get_api_from_relative_path(
-        response.raise_for_status()
-        # raise for status only handles: 400 <= status_code < 600
-        raise Exception(f"response.status_code: {response.status_code}", response.text)
+        if response:
+            response.raise_for_status()
+            # raise for status only handles: 400 <= status_code < 600
+            msg = f"Unexpected status {response.status_code} for url: {response.url}"
+            raise HTTPError(msg, response=response)
+        msg = "Missing response object in _get_api_from_relative_path"
+        raise SystemError(msg)
```
I _have_ seen this in our logs, where the `response.raise_for_status()` itself throws an error, but it's rare so I'm not sure the real-life circumstances that cause it.

jbristow commented on nodestream_akamai/akamai_utils/cloudlet_client.py:
```diff
@@ -1,7 +1,5 @@
 import logging
-import re
 from typing import List
-from urllib.parse import urlparse
```
optimize imports

jbristow commented on nodestream_akamai/akamai_utils/cloudlets_v2_client.py:
```diff
@@ -9,7 +9,7 @@
 
 
 class AkamaiCloudletsV2Client(AkamaiApiClient):
-    def cloudlet_policy_ids(self) -> List[int]:
```
type mismatch

jbristow commented on nodestream_akamai/akamai_utils/cloudlets_v2_client.py:
```diff
@@ -26,7 +26,7 @@ def cloudlet_policy_ids(self) -> List[int]:
 
         return {policy["policyId"] for policy in returned_policies}
 
-    def cloudlet_policy_ids_er(self) -> List[int]:
```
type mismatch

jbristow commented on nodestream_akamai/akamai_utils/model.py:
```diff
@@ -10,24 +10,26 @@ class EdgeHost:
     hostnames: List[EdgeHost]
-    ruleFormat: str
-    deeplink: str
+    version: str | None = None
+    siteshield_maps: List[str] = field(default_factory=list)
+    rule_format: str | None = None
+    image_manager_policysets: List[str] = field(default_factory=list)
+    edgeworker_ids: List[int] = field(default_factory=list)
+    edge_redirector_policies: List[int] = field(default_factory=list)
+    deeplink: str | None = None
+    cloudlet_policies: List[int] = field(default_factory=list)
+    cp_codes: List[int] = field(default_factory=list)
```
without defaults, pycharm was warning that required constructor arguments were missing.

jbristow commented on nodestream_akamai/property/property.py:
```diff
@@ -2,39 +2,36 @@
         except Exception as err:
-            self.logger.error("Failed to list properties: %s", err)
-            return
+            self.logger.exception("Failed to list properties: %s", err)
+            raise err
```
`logger.exception("msg")` is equivalent to `logger.error("msg", exc_info=True)` and is the preferred way to include stack trace and other info into the log.

jbristow commented on nodestream_akamai/akamai_utils/property_client.py:
```diff
@@ -130,19 +208,19 @@ def describe_property_by_id(self, property_id: str) -> PropertyDescription:
         )
 
-    def describe_property_by_dict(self, property: dict) -> PropertyDescription:
+    def describe_property_by_dict(self, prop: dict) -> PropertyDescription:
```
`property` is a python builtin, and I find it best to avoid using names that shadow them

jbristow commented on nodestream_akamai/akamai_utils/property_client.py:
```diff
@@ -222,292 +300,163 @@ def list_all_properties(self):
+        for match_type in MATCH_TYPES:
             criteria_results[match_type] = []
-            criterion_results = {}
+            criterion_results = {k: [] for k in MATCH_TYPES}
```
sometimes the inner loops were trying to use the keys before they were added to the dict

jbristow commented on nodestream_akamai/akamai_utils/property_client.py:
```diff
@@ -530,15 +479,16 @@ def parse_origin_location(self, rule_location, parent_location, rules):
 
         return location_results
```
Is this updated anywhere? Is the code in-between actually doing anything?

jbristow commented on nodestream_akamai/akamai_utils/property_client.py:
```diff
@@ -222,292 +300,163 @@ def list_all_properties(self):
+            location_results = self.parse_origin_location(
+                location, parent_location, rules
+            )
+            if len(location_results.get("path", [])) > 0:
+                combined_rule_paths.append(location_results["path"])
+            if len(location_results.get("hostname", [])) > 0:
+                combined_rule_hosts.append(location_results["hostname"])
+            if len(location_results.get("cloudOrigin", [])) > 0:
+                combined_rule_cdids.append(location_results["cloudOrigin"])
```
these keys were sometimes not in the `location_results` and threw KeyErrors
